### ISSUE TITLE
Rehydration causes memory spikes equal to source size

### ISSUE THREAD

**Issue by ggnall** [MaterializeInc/materialize#19866]
We have this unfortunate scenario that can happen/is likely to happen:
• User installs a materialized view on a source, with an dataflow that doesn’t have any joins or reduces, so no significant arrangements get installed. The data mostly just flows from persist into persist.
• The source accumulates data over time. The cumulative data size grows larger than the memory size of the replica. This is fine, because the data arrives distributed over time, so not much of it must be kept in memory at a time.
• The replica is restarted, for example during a weekly update.
• The MVs dataflow reads the data from the source again, which is now compacted to a single time.
• persist_sink (and possibly other operators) collects data in memory, as long as the input frontier doesn’t advance. This means it will attempt to keep the whole contents of the source in memory, but the memory is not large enough for that.
• The replica OOMs, tries to rehydrate the dataflow again, OOMs again, … 

[Slack Message](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1668799218878549?thread_ts=1668799218.878549&cid=C02PPB50ZHS)

**Comment by ggnall** [comment]
@chuck-alt-delete: even (especially) stateless materialized views cause memory spikes equal to the result set on rehydration. That means if you are parsing json, using no memory, on a 100GB source, rehydration will spike memory to 100GB before settling back to ~0. 

This prevents users from scaling down clusters after initial ingestion (or will cause issues if they do so). Current approach has been to advise users to rely on clusters that can maintain a Materialized Views snapshot size. 

From https://materializeinc.slack.com/archives/C02PPB50ZHS/p1668799218878549 this seems to span persist and compute cc @teskje @danhhz

**Comment by teskje** [comment]
This is a relevant comment from Dan: https://github.com/MaterializeInc/materialize/issues/15974#issuecomment-1587575061

**Comment by teskje** [comment]
A while ago @aljoscha wrote a doc summarizing the issue: [in Notion](https://www.notion.so/materialize/Smooth-Re-hydration-464bbbf76ed640e78a5948c49b0e6ae4). Specifically, the "Many updates at one/the same timestamp" part is relevant here.

**Comment by teskje** [comment]
One thing that we could try is to implement "targeted swapping" for persist_sink during hydration using mmap. We would only resort to using an mmaped file while the persist_sink is still hydrating, and probably only once some size threshold is reached. This will hopefully mitigate the negative effect using mmap has on performance sufficiently.

[Relevant Slack thread.](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1695286995544919)

**Comment by teskje** [comment]
Closing in favor of https://github.com/MaterializeInc/materialize/issues/19866 which has a more up-to-date description of the problem, and a more findable title.
